### ISSUE TITLE
Chords within a chord don't work when created with a group and a pipe but work when created with the chord function

### ISSUE THREAD

**Issue by Rubinous** [celery/celery#2922]
By looking at some other issue reports, there seems to be an eluding bug related to chords within chords. However, I was accidentally able to find a way to build the chords so that the bug appears and so that it doesn't appear. This should make the issue clear and easy to debug.

This works:
```
In [18]: inner_chord1 = chord([mul.s(2, 3), mul.s(3, 4)], xsum.s())

In [19]: inner_chord2 = chord([mul.s(4, 5), mul.s(5, 6)], xsum.s())

In [20]: outer_chord = group(inner_chord1, inner_chord2) | xsum.s()

In [21]: outer_chord()
```
The last task executes correctly and returns with `68`.

But if I create the inner chords with a group and a pipe:
```
In [22]: bugged_inner_chord1 = group(mul.s(2, 3), mul.s(3, 4)) | xsum.s()

In [23]: bugged_inner_chord2 = group(mul.s(4, 5), mul.s(5, 6)) | xsum.s()

In [24]: bugged_outer_chord = group(bugged_inner_chord1, bugged_inner_chord2) | xsum.s()

In [25]: bugged_outer_chord()
```
the last `xsum` never gets executed!!!

Code for the tasks:
```
@task(ignore_result=False)
def xsum(numbers):
    return sum(numbers)


@task(ignore_result=False)
def mul(a, b):
    return a * b
```

Using Redis as a result backend with `CELERY_IGNORE_RESULT = False`. Using the same Redis instance as the broker.

Versions: amqp-1.4.7 billiard-3.3.0.21 celery-3.1.19 kombu-3.0.29 and Redis server v=3.0.4.

**Comment by mes5k** [comment]
Hmmm, with 4.0.0 rc1 the "good" solution above returns 18, although I see both inner chords calculating the correct results in the logs.  

The "buggy" solution fails outright with this

    Traceback (most recent call last):
     File "other_run.py", line 20, in <module>
        badish()
      File "other_run.py", line 16, in badish
        res = bugged_outer_chord.delay()
      File "/Library/Python/2.7/site-packages/celery-4.0.0rc1-py2.7.egg/celery/canvas.py", line 188, in delay
        return self.apply_async(partial_args, partial_kwargs)
      File "/Library/Python/2.7/site-packages/celery-4.0.0rc1-py2.7.egg/celery/canvas.py", line 408, in     apply_async
        dict(self.options, **options) if options else self.options))
      File "/Library/Python/2.7/site-packages/celery-4.0.0rc1-py2.7.egg/celery/canvas.py", line 431, in run
        chain=tasks if not use_link else None, **options)
      File "/Library/Python/2.7/site-packages/celery-4.0.0rc1-py2.7.egg/celery/canvas.py", line 919, in     apply_async
        return self.run(tasks, body, args, task_id=task_id, **options)
      File "/Library/Python/2.7/site-packages/celery-4.0.0rc1-py2.7.egg/celery/canvas.py", line 961, in run
        result=results)
      File "/Library/Python/2.7/site-packages/celery-4.0.0rc1-py2.7.egg/celery/backends/base.py", line 554, in _apply_chord_incr
        return header(*partial_args, task_id=group_id, **fixed_options or {})
      File "/Library/Python/2.7/site-packages/celery-4.0.0rc1-py2.7.egg/celery/canvas.py", line 787, in __call__
        return self.apply_async(partial_args, **options)
      File "/Library/Python/2.7/site-packages/celery-4.0.0rc1-py2.7.egg/celery/canvas.py", line 746, in apply_async
        group_id, list(self._apply_tasks(tasks, producer, app, **options)),
      File "/Library/Python/2.7/site-packages/celery-4.0.0rc1-py2.7.egg/celery/canvas.py", line 720, in _apply_tasks
        **options)
      File "/Library/Python/2.7/site-packages/celery-4.0.0rc1-py2.7.egg/celery/canvas.py", line 408, in    apply_async
        dict(self.options, **options) if options else self.options))
      File "/Library/Python/2.7/site-packages/celery-4.0.0rc1-py2.7.egg/celery/canvas.py", line 431, in run
        chain=tasks if not use_link else None, **options)
    TypeError: apply_async() got multiple values for keyword argument 'chain'


Versions: 
    amqp (2.0.0rc1)
    billiard (3.4.0.0b1)
    celery (4.0.0rc1)
    kombu (4.0.0a1)

**Comment by brakhane** [comment]
@mes5k JFYI, the apply_async() bug is already tracked as #2913

**Comment by ask** [comment]
I fixed the apply bug, so both of these tests now correctly returns 68 in master

But that's only for the chord unlock task backends, redis returns 18 for some reason.

**Comment by Rubinous** [comment]
@ask, what's a chord unlock task backend? This is still relevant to me. Would I be safe with some other result backend than Redis?

**Comment by ask** [comment]
After I made the comment I pushed a change that also fixes this for the Redis result backend (649f616)

That would only be in the master branch though, not 3.1
